Stratis.Bitcoin/BlockPulling/BlockPuller.cs

```diff
@@ -448,5 +356,279 @@ private void DistributeDownload(Dictionary<int, InventoryVector> vectors, BlockP
+        /// <param name="blockHash">Hash of the block which task should be released.</param>
+        /// <returns><c>true</c> if the function succeeds, <c>false</c> if the block was not assigned to be downloaded by any peer.</returns>
+        /// <exception cref="InvalidOperationException">Thrown in case of data inconsistency between synchronized structures, which should never happen.</exception>
+        internal bool ReleaseDownloadTaskAssignment(BlockPullerBehavior peer, uint256 blockHash)
+        {
+            bool res = false;
+            lock (this.lockObject)
+            {
+                HashSet<uint256> peerPendingDownloads;
+                if (this.pendingDownloads.TryGetValue(peer, out peerPendingDownloads))
+                    res = ReleaseDownloadTaskAssignmentLocked(peerPendingDownloads, blockHash);
+            }
+
+            if (!res) throw new InvalidOperationException("Data structures inconsistency, please notify the devs");
+
+            return res;
+        }
+
+        /// <summary>
+        /// Releases the block downloading task from the peer it has been assigned to 
+        /// and returns the block to the list of blocks the node wants to download.
+        /// </summary>
+        /// <param name="peerPendingDownloads">List of pending downloads tasks of the peer.</param>
+        /// <param name="blockHash">Hash of the block which task should be released.</param>
+        /// <returns><c>true</c> if the function succeeds, <c>false</c> if the block was not assigned to be downloaded by any peer.</returns>
+        /// <remarks>The caller of this method is responsible for holding <see cref="lockObject"/>.</remarks>
+        private bool ReleaseDownloadTaskAssignmentLocked(HashSet<uint256> peerPendingDownloads, uint256 blockHash)
+        {
+            bool res = false;
+
+            if (this.map.Remove(blockHash) && peerPendingDownloads.Remove(blockHash))
+            {
+                this.pendingInventoryVectors.Enqueue(blockHash);
+                res = true;
+            }
+
+            return res;
+        }
+
+        /// <summary>
+        /// Releases all pending block download tasks assigned to a peer.
+        /// </summary>
+        /// <param name="peer">Peer to have all its pending download task released.</param>
+        /// <exception cref="InvalidOperationException">Thrown in case of data inconsistency between synchronized structures, which should never happen.</exception>
+        internal void ReleaseAllPeerDownloadTaskAssignments(BlockPullerBehavior peer)
+        {
+            lock (this.lockObject)
+            {
+                HashSet<uint256> peerPendingDownloads;
+                if (this.pendingDownloads.TryGetValue(peer, out peerPendingDownloads))
+                {
+                    // Make a fresh copy of items in peerPendingDownloads to avoid modification of the collection.
+                    foreach (uint256 blockHash in peerPendingDownloads.ToList())
```

Comment:
Ah yeah I see my mistake peerPendingDownloads is passed untouched and ReleaseDownloadTaskAssignmentLocked will remove from it.